src/main/java/io/nats/client/impl/Headers.java

```diff
@@ -0,0 +1,217 @@
+	 */
+	public Headers put(String key, String... values) {
+		if (values != null) {
+			put(key, Arrays.asList(values));
+		}
+		return this;
+	}
+
+	/**
+	 * Associates the specified values with the key. If the key was already present
+	 * any existing values are removed and replaced with the new list.
+	 * null values are ignored. If all values are null, the put is ignored
+	 *
+	 * @param key the key
+	 * @param values the values
+	 * @throws IllegalArgumentException if the key is null or empty or contains invalid characters
+	 *         -or- if any value contains invalid characters
+	 */
+	public Headers put(String key, Collection<String> values) {
+		if (values != null) {
+			List<String> checked = checkKeyValues(key, values);
+			if (!checked.isEmpty()) {
+				headerMap.put(key, checked);
+				valueCsvCache.remove(key); // remove from cache so it will be re-calculated
+			}
+		}
+		return this;
+	}
+
+	/**
+	 * Removes each key and its values if the key was present
+	 *
+	 * @param keys the key or keys to remove
+	 */
+	public void remove(String... keys) {
+		for (String key : keys) {
+			headerMap.remove(key);
+			valueCsvCache.remove(key); // remove from cache so it will be re-calculated
+		}
+	}
+
+	/**
+	 * Removes each key and its values if the key was present
+	 *
+	 * @param keys the key or keys to remove
+	 */
+	public void remove(Collection<String> keys) {
+		for (String key : keys) {
+			headerMap.remove(key);
+			valueCsvCache.remove(key); // remove from cache so it will be re-calculated
+		}
+	}
+
+	public int size() {
+		return headerMap.size();
+	}
+
+	public boolean isEmpty() {
+		return headerMap.isEmpty();
+	}
+
+	public void clear() {
+		headerMap.clear();
+	}
+
+	public boolean containsKey(String key) {
+		return headerMap.containsKey(key);
+	}
+
+	public List<String> values(String key) {
+		List<String> list = headerMap.get(key);
+		return list == null ? null : Collections.unmodifiableList(list);
+	}
+
+	public Set<String> keySet() {
+		return headerMap.keySet();
+	}
+
+	private List<String> checkKeyValues(String key, Collection<String> values) {
+		checkKey(key);
+
+		List<String> checked = new ArrayList<>();
+		if (values != null && !values.isEmpty()) {
+			for (String v : values) {
+				if ( checkValue(v) ) {
+					checked.add(v);
+				}
+			}
+		}
+		return checked;
+	}
+
+	private void checkKey(String key) {
+		// key cannot be null or empty and contain only printable characters except colon
+		if (key == null || key.length() == 0) {
+			throw new IllegalArgumentException(KEY_CANNOT_BE_EMPTY_OR_NULL);
+		}
+
+		key.chars().forEach(c -> {
+			if (c < 32 || c > 126 || c == ':') {
+				throw new IllegalArgumentException(KEY_INVALID_CHARACTER + c);
+			}
+		});
+	}
+
+	private boolean checkValue(String val) {
+		// Generally more permissive than HTTP.  Allow only printable
+		// characters and include tab (0x9) to cover what's allowed
+		// in quoted strings and comments.
+		// null is just ignored
```

Comment:
Since this operates on single ASCII bytes we should be able to avoid strings/chars here I think.